mcs/class/System/System.Net.Sockets/SocketAsyncResult.cs

```diff
@@ -155,7 +155,11 @@ public void Complete ()
 
 			AsyncCallback callback = AsyncCallback;
 			if (callback != null) {
-				ThreadPool.UnsafeQueueUserWorkItem (_ => callback (this), null);
+				QueueUserWorkItemCallback workItemCallback = (object state) => {
```

Comment:
Could you replace that with `ThreadPool.UnsafeQueueUserWorkItem(state => ((SocketAsyncResult)state).AsyncCallback(state), this)`? And if you want to reduce allocations even further, move this lambda to a static delegate and pass this static delegate to `ThreadPool.UnsafeQueueUserWorkItem`